src/sql/src/plan/statement/ddl.rs

```diff
@@ -911,7 +916,55 @@ pub fn plan_create_source(
             if key_envelope == KeyEnvelope::None {
                 key_envelope = get_unnamed_key_envelope(key_encoding)?;
             }
-            UnplannedSourceEnvelope::Upsert(UpsertStyle::Default(key_envelope))
+
+            let order_by = if !order_by.is_empty() {
+                scx.require_unsafe_mode("ENVELOPE UPSERT ORDER BY")?;
+
+                let mut offset_included = false;
+
+                let metadata_indices = order_by
+                    .iter()
+                    .map(|order_col| {
+                        let metadata_type = metadata_columns
+                            .iter()
+                            .position(|(col_name, _)| *col_name == order_col.as_str())
+                            .map(|pos| {
+                                let included_col = metadata_column_types
+                                    .get(pos)
+                                    .expect("metadata index calculation error");
+                                (pos, included_col)
+                            });
+
+                        match metadata_type {
+                            Some((pos, &IncludedColumnSource::Timestamp)) => Ok(pos),
+                            Some((pos, &IncludedColumnSource::Offset)) => {
+                                offset_included = true;
+                                Ok(pos)
+                            }
+                            Some(_) => sql_bail!(
+                                "Only TIMESTAMP and OFFSET are supported as order by columns"
+                            ),
+                            None => sql_bail!("Could not find {} in included metadata", order_col),
```

Comment:
`OFFSET` has to be last right? do we not check this?

Reply:
Also, in a comment here can we note that `ORDER BY OFFSET` is implicit when you don't have an `ORDER BY`, but also _when you dont specify an `ORDER BY`, you don't need to also `INCLUDE OFFSET`_

Reply:
@guswynn Technically `OFFSET` can be in the front or middle even. Just that any order by columns coming after offset will be meaningless in practice.